core/src/main/scala/com/gu/tableversions/core/InMemoryTableVersions.scala

```diff
@@ -1,16 +1,102 @@
+    extends TableVersions[F] {
+
+  override def init(table: TableName): F[Unit] =
+    allUpdates.update { prev =>
+      if (prev.contains(table)) prev else prev + (table -> Nil)
+    }
+
+  override def currentVersion(table: TableName): F[TableVersion] =
+    // Derive current version of a table by folding over the history of changes
+    for {
+      allTableUpdates <- allUpdates.get
+      tableUpdates <- allTableUpdates
+        .get(table)
+        .fold(F.raiseError[List[TableUpdate]](new Exception(s"Table '$table' not found")))(F.pure)
+      operations = tableUpdates.flatMap(_.partitionUpdates)
+    } yield InMemoryTableVersions.applyUpdate(TableVersion.empty)(operations)
+
+  override def nextVersions(table: TableName, partitions: List[Partition]): F[List[PartitionVersion]] = {
+    def incrementVersion(partitionVersion: PartitionVersion): PartitionVersion =
+      partitionVersion.copy(version = VersionNumber(partitionVersion.version.number + 1))
+
+    for {
+      tableVersion <- currentVersion(table)
+      nextVersions = partitions.map(
+        partition =>
+          tableVersion.partitionVersions
+            .find(_.partition == partition)
+            .map(incrementVersion)
+            .getOrElse(PartitionVersion(partition, VersionNumber(1))))
+    } yield nextVersions
+  }
+
+  override def commit(table: TableName, update: TableVersions.TableUpdate): F[TableVersions.CommitResult] = {
+
+    // Note: we're not checking invalid partition versions here as we suspect this problem will go
+    // away in a later iteration of this interface.
+
+    val applyUpdate: TableUpdates => Either[Exception, TableUpdates] = { currentTableUpdates =>
+      if (currentTableUpdates.contains(table)) {
+        val updated = currentTableUpdates + (table -> (currentTableUpdates(table) :+ update))
+        Right(updated)
+      } else
+        Left(new Exception(s"Unknown table '${table.fullyQualifiedName}'"))
+    }
+
+    allUpdates.modifyEither(applyUpdate).map(_ => SuccessfulCommit)
+  }
+
+}
+
+object InMemoryTableVersions {
+
+  type TableUpdates = Map[TableName, List[TableUpdate]]
+
+  /**
+    * Safe constructor
+    */
+  def apply[F[_]](implicit F: Sync[F]): F[InMemoryTableVersions[F]] =
+    for {
+      ref <- Ref[F].of(Map[TableName, List[TableUpdate]]())
+      service = new InMemoryTableVersions[F](ref)
+    } yield service
+
+  /**
+    * Produce current table version based on history of updates.
+    */
+  private[core] def applyUpdate(initial: TableVersion)(operations: List[PartitionOperation]): TableVersion = {
+
+    def applyOp(agg: Map[Partition, VersionNumber], op: PartitionOperation): Map[Partition, VersionNumber] = op match {
+      case AddPartitionVersion(partitionVersion: PartitionVersion) =>
+        agg + (partitionVersion.partition -> partitionVersion.version)
+      case RemovePartition(partition: Partition) =>
+        agg - partition
+    }
```

Comment:
I have a feeling all of this could be simplified using, umm, monoids? 😁 Or some other clever Cats bits. Suggestions welcome!